core/src/main/java/org/infinispan/stream/impl/DistributedCacheStream.java

```diff
@@ -0,0 +1,803 @@
+      if (iteratorOperation != IteratorOperation.FLAT_MAP) {
+         iteratorOperation = IteratorOperation.MAP;
+      }
+      return addIntermediateOperationMap(new MapToDoubleOperation<>(mapper), doubleCacheStream());
+   }
+
+   @Override
+   public <R1> Stream<R1> flatMap(Function<? super R, ? extends Stream<? extends R1>> mapper) {
+      iteratorOperation = IteratorOperation.FLAT_MAP;
+      return addIntermediateOperationMap(new FlatMapOperation<>(mapper), (Stream<R1>) this);
+   }
+
+   @Override
+   public IntStream flatMapToInt(Function<? super R, ? extends IntStream> mapper) {
+      iteratorOperation = IteratorOperation.FLAT_MAP;
+      return addIntermediateOperationMap(new FlatMapToIntOperation<>(mapper), intCacheStream());
+   }
+
+   @Override
+   public LongStream flatMapToLong(Function<? super R, ? extends LongStream> mapper) {
+      iteratorOperation = IteratorOperation.FLAT_MAP;
+      return addIntermediateOperationMap(new FlatMapToLongOperation<>(mapper), longCacheStream());
+   }
+
+   @Override
+   public DoubleStream flatMapToDouble(Function<? super R, ? extends DoubleStream> mapper) {
+      iteratorOperation = IteratorOperation.FLAT_MAP;
+      return addIntermediateOperationMap(new FlatMapToDoubleOperation<>(mapper), doubleCacheStream());
+   }
+
+   @Override
+   public Stream<R> distinct() {
+      DistinctOperation op = DistinctOperation.getInstance();
+      markDistinct(op, IntermediateType.OBJ);
+      return addIntermediateOperation(op);
+   }
+
+   @Override
+   public Stream<R> sorted() {
+      markSorted(IntermediateType.OBJ);
+      return addIntermediateOperation(SortedOperation.getInstance());
+   }
+
+   @Override
+   public Stream<R> sorted(Comparator<? super R> comparator) {
+      markSorted(IntermediateType.OBJ);
+      return addIntermediateOperation(new SortedComparatorOperation<>(comparator));
+   }
+
+   @Override
+   public Stream<R> peek(Consumer<? super R> action) {
+      return addIntermediateOperation(new PeekOperation<>(action));
+   }
+
+   @Override
+   public Stream<R> limit(long maxSize) {
+      LimitOperation op = new LimitOperation<>(maxSize);
+      markDistinct(op, IntermediateType.OBJ);
+      return addIntermediateOperation(op);
+   }
+
+   @Override
+   public Stream<R> skip(long n) {
+      SkipOperation op = new SkipOperation<>(n);
+      markSkip(IntermediateType.OBJ);
+      return addIntermediateOperation(op);
+   }
+
+   // Now we have terminal operators
+
+   @Override
+   public R reduce(R identity, BinaryOperator<R> accumulator) {
+      return performOperation(TerminalFunctions.reduceFunction(identity, accumulator), true, accumulator, null);
+   }
+
+   @Override
+   public Optional<R> reduce(BinaryOperator<R> accumulator) {
+      R value = performOperation(TerminalFunctions.reduceFunction(accumulator), true,
+              (e1, e2) -> {
+                 if (e1 != null) {
+                    if (e2 != null) {
```

Comment:
Yes, it is possible the reduce operated on an empty stream which in turn would return null  If you notice I specifically wrote this to have to serialize the Optional instance.